src/libbson/tests/test-bson-vector.c

```diff
@@ -0,0 +1,1400 @@
+{
+   // (TODO for DRIVERS-3095, DRIVERS-3097) This implements something the test covers that our API doesn't. If the test
+   // were modified to cover element-by-element conversion, this can be replaced with
+   // bson_append_vector_packed_bits_from_array.
+
+   BSON_ASSERT_PARAM (bson);
+   BSON_ASSERT_PARAM (key);
+   BSON_ASSERT_PARAM (iter);
+
+   size_t byte_count = 0;
+   {
+      bson_iter_t validation_iter = *iter;
+      while (bson_iter_next (&validation_iter)) {
+         if (!BSON_ITER_HOLDS_INT (&validation_iter)) {
+            bson_set_error (error,
+                            BSON_ERROR_VECTOR,
+                            BSON_VECTOR_ERROR_ARRAY_ELEMENT_TYPE,
+                            "expected int32 or int64 in BSON array key '%s', found item type 0x%02X",
+                            bson_iter_key (&validation_iter),
+                            (unsigned) bson_iter_type (&validation_iter));
+            return false;
+         }
+         int64_t byte_as_int64 = bson_iter_as_int64 (&validation_iter);
+         if (byte_as_int64 < 0 || byte_as_int64 > UINT8_MAX) {
+            bson_set_error (error,
+                            BSON_ERROR_VECTOR,
+                            BSON_VECTOR_ERROR_ARRAY_ELEMENT_VALUE,
+                            "BSON array key '%s' value %" PRId64 " is out of range for packed byte",
+                            bson_iter_key (&validation_iter),
+                            byte_as_int64);
+            return false;
+         }
+         byte_count++;
+      }
+   }
+
+   if (padding < 0 || padding > 7) {
+      bson_set_error (error,
+                      TEST_ERROR_DOMAIN,
+                      TEST_ERROR_CODE,
+                      "'padding' parameter (%" PRId64
+                      ") for append_vector_packed_bits_from_packed_array is out of range",
+                      padding);
+      return false;
+   }
+   if (byte_count < 1 && padding > 0) {
+      bson_set_error (error,
+                      TEST_ERROR_DOMAIN,
+                      TEST_ERROR_CODE,
+                      "nonzero 'padding' parameter (%" PRId64
+                      ") for zero-length append_vector_packed_bits_from_packed_array",
+                      padding);
+      return false;
+   }
+
+   bson_vector_packed_bits_view_t view;
+   if (bson_append_vector_packed_bits (bson, key, key_length, byte_count * 8 - padding, &view)) {
+      bson_iter_t copy_iter = *iter;
+      for (size_t i = 0; i < byte_count; i++) {
+         BSON_ASSERT (bson_iter_next (&copy_iter));
+         uint8_t packed_byte = (uint8_t) bson_iter_as_int64 (&copy_iter);
+         BSON_ASSERT (bson_vector_packed_bits_view_write_packed (view, &packed_byte, 1, i));
+      }
+      return true;
+   } else {
+      return false;
+   }
+}
+
+static void
+hex_str_to_bson (bson_t *bson_out, const char *hex_str)
+{
+   uint32_t size = strlen (hex_str) / 2;
```

Comment:
```suggestion
   uint32_t size = (uint32_t) strlen (hex_str) / 2u;
```
Addresses `-Wshorten-64-to-32` warnings.